### PR TITLE
Validate install_origins

### PR DESCRIPTION
**Pull request by diox**
Fixes https://github.com/mozilla/addons-linter/issues/3972

### CODE REVIEW

File: src/schema/formats.js
Comment (rpl) [inline src/schema/formats.js:134]: We should keep this as much as possible in sync with what we are going to do on the Firefox side, when the patch on the Firefox side will be landed it may be a good idea to put a permalink to the formatter used in Firefox.

In the meantime, here is the related review comment I added on the Firefox patch:
https://phabricator.services.mozilla.com/D129105#inline-715951

Shane and I haven't touched base about that yet, but on the Firefox side we will proceed to use a Schemas formatter like the one I described in that phabricator comment, then in nodejs I think that it may look more or less like the following code snippet:
```js
const ALLOWED_INSTALL_ORIGINS_PROTOCOLS = ["http:", "https:"];

...

function isValidInstallOriginUrl(value) {
  let origin;
  try {
    const url = new URL(value);
    if (!ALLOWED_INSTALL_ORIGINS_PROTOCOLS.includes(url.protocol)) {
      // URL protocol is not one of the allowed ones (e.g. ["http:", "https:"]).
      return false;
    }
    origin = url.origin;
  } catch (err) {
    // Not a valid url.
    return false;
  }

  const { href: baseURL } = new URL("/", origin);
  // Make sure value URL doesn't include any other URL component besides the origin
  // and path set to "/". 
  return baseURL.slice(0, -1) === value;
} 
```

File: src/schema/imported/manifest.json
Comment (rpl) [inline src/schema/imported/manifest.json:792]: As a side note, as part of the review on the Firefox patch, I did mention that `FirefoxSpecificProperties` JSONSchema doesn't allow arbitrary properties that are not explicitly listed in its schema, and so a manifest.json file including `install_origins` inside `browser_specific_settings.gecko` (or `applications.gecko`) would be invalid for older Firefox releases (summarized from [this comment](https://phabricator.services.mozilla.com/D129105#inline-715990)).

I briefly touched base about that with Shane yesterday, and we agreed to discuss about it more today during out team standup.

And so, for now there are no changes requested, but we may have to change this to match what we will agree on doing on the Firefox side (e.g. we can't exclude yet that we may choose to move the property outside of `FirefoxSpecificProperties`, e.g. in a new top level manifest key where the pre-existing JSONSchema does already allow additional properties).

File: tests/unit/parsers/test.manifestjson.js
Comment (rpl) [inline tests/unit/parsers/test.manifestjson.js:3642]: Nit, we may use `it.each` Jest helper as we do for some other similar test cases, e.g. this one related to the homepage_url:
https://github.com/mozilla/addons-linter/blob/563ad4ffa2cb877fbc94c29debf5a3759aad18a9/tests/unit/parsers/test.manifestjson.js#L2977-L2982

File: tests/unit/parsers/test.manifestjson.js
Comment (rpl) [inline tests/unit/parsers/test.manifestjson.js:3637]: Side note: it would be nice to somehow make it easier for us to keep at least some of the test cases that we will have to run on both Firefox and the addons-linter sides in sync between them.

Sharing the entire test case is not really doable, but may be worth moving the list of test case into arrays (in both the Firefox tests and these tests) and just make sure the list is as in sync as it can be (especially in terms of cases being tested).

**Review by rpl — COMMENTED** [review]
@diox my apologies for letting you wait that long. 

This PR seems to be defintiely going in the right direction.

Follows a more detailed initial round of feedback comments (included a mention to some details that we may still have to discuss and agree on for the Firefox side of this change, which may require changes to this patch, e.g. if we will end up moving the `install_origins` manifest property).

File: tests/unit/parsers/test.manifestjson.js
Comment (diox) [reply tests/unit/parsers/test.manifestjson.js:3637]: Yeah. Kinda similar to your `it.each` suggestion below, the issue with having arrays containing a bunch of invalid/valid values is that when it's invalid, the reason can be different, and we want to check it... Not sure if it's worth doing several arrays of invalid values, one for each kind of error...

File: src/schema/imported/manifest.json
Comment (rpl) [reply src/schema/imported/manifest.json:792]: > TIL. We want developers to adopt `install_origins` sooner rather than later, if it makes their add-on incompatible with older Firefox versions that sounds like a significant problem...

It is definitely a significant problem (and I'm really glad that we spotted it earlier in the review passes).
That's basically why yesterday I asked Shane to have a quick preliminary chat about that single point from my review pass on the Firefox patch, and we agreed to discuss about it more during the team standup today.

I'm confident we will get to an agreement quickly, either today or in the worst case by the end of this week (you are also welcome to join us if you want to, I can ping you when we are about to discuss about this point, but in any case I will report back here the outcome of that discussion).

Comment (rpl) [reply src/schema/imported/manifest.json:792]: @diox we discussed about this issue during the meeting (and Shane told me that it has been discussed also in another meeting, I think it was the Ops one) and we agreed to move the key to be a top level manifest key (same name and same JSONSchema definition, just moved to be a top level one).

I'm not yet sure if it is meant to be on the `ManifestBase` one, but it seems a reasonable assumption given that it was in the `FirefoxSpecificProperties`, which are part of the `ManifestBase` schema.

File: tests/unit/parsers/test.manifestjson.js
Comment (willdurand) [reply tests/unit/parsers/test.manifestjson.js:3642]: nit: also if we can avoid `testXXX` and have the assertions directly in the test function, that would avoid some ESLint warnings about tests without assertions :)

File: tests/unit/parsers/test.manifestjson.js
Comment (willdurand) [reply tests/unit/parsers/test.manifestjson.js:3637]: Not sure exactly if that's what you're looking for but we can totally do:

```
it.each([
  { origin: 'xyz', expectedError: /some error/ },
  { origin: 'qbc', expectedError: /some other error/ },
])() => {
});
```

Comment (diox) [reply tests/unit/parsers/test.manifestjson.js:3637]: Sure, I can do that. Note that if we want to share the test values then Firefox should also have the expected error code in its tests.

File: src/schema/imported/manifest.json
Comment (rpl) [inline src/schema/imported/manifest.json:784]: @diox given that it looks that we now got to an agreement about the pending issue related to this property [^1], and that I'm already reviewing the patches on the Firefox side, I though to run the schema imports for you and push them into my fork (in a [branch named `pre-review/install-origins`](https://github.com/mozilla/addons-linter/compare/master...rpl:pre-review/install-origins)) to make it easier for you to cherry pick what is needed directly from there, the one that may be more directly useful for your work in this pull request are:
- https://github.com/mozilla/addons-linter/commit/b94bba48e4559fdba9deb26293eb359da3d8d859 (this one is just the schema import of the "install_origins" definition from [D129105](https://phabricator.services.mozilla.com/D129105), other schema unrelated to `install_origins' is in separate patches in that same fork, hopefully you will only need this one to try our your changes)
- https://github.com/mozilla/addons-linter/commit/720066a79a03bdb64d13b1274cc5084d27bca927 (this one contains the small tweak that we will need to add to `src/schema/updates/manifest.json`, plus the changes to `src/schema/updates/manifest.json` got by re-importing the schema again on top of the current version of this patch, which means that it does remove this definition and instead adds "maxItems" to the one added by the schema data import from the other patch + a small tweak to the test case to confirm that everything else works as expected)

Let me know how it goes when you will give it a look to cherry picking what's needed from that branch (especially if there is something that isn't really clear or if it isn't working for you).

I'm going to mark the other comment (https://github.com/mozilla/addons-linter/pull/4018/files/0d3278bc1b2fbaad5431df465d19930830757ac3#r745572036) as resolved.

[^1]: in particular: where the new property should be placed (top level manifest property in the `ManifestBase` schema), its manifest property name (`install_origins`) and what should be the format expected to be enforced (`origin` as defined by the specs), I did

**Comment by rpl** [comment]
(as a side note: I added the label "state: blocked by upstream" as a reminder that this may not be merged as is until the schema changes on the Firefox side got to beta even if we have signed it off already in the meantime, we can evaluate if it is possible and we want/need to split out part of this and landing sooner later on, after the PR is ready or almost ready from the reviewers perspective).

File: src/schema/imported/manifest.json
Comment (rpl) [reply src/schema/imported/manifest.json:784]: @diox here is a patch for the strategy that we have been discussing over Matrix today (I did rerun the import from the current beta on top of the change to the updates/manifest.json file, and tweaked the test case from your PR to confirm that they all pass when the install_origins is moved to the top level):
https://github.com/mozilla/addons-linter/commit/3df53d6fc6802aae252b57266620d9f448119127

**Comment by rpl** [comment]
I'm going to remove the `state: blocked by upstream` label, the approach we discussed this morning ([^1]), and drafted in the patch linked in https://github.com/mozilla/addons-linter/pull/4018#discussion_r751194894, will allow us to don't block merging this once signed-off.

[^1]: "injecting the entire "install_origin" schema entry from src/updates/manifest.json" will allow us to keep the entry in place on schema data imports that we may have to do before the Firefox release with the same schema data will actually get to beta.

File: src/schema/formats.js
Comment (rpl) [inline src/schema/formats.js:118]: We may not need to explicitly check for "*" given that this is an origin and we are already going to require that the string is strictly equal to the origin (and with that check we should already excluding a number of url components where `new URL(...)` would not be complaining about a "*"), that should be leaving only protocol, hostname and protocol and new URL should be throwing already with a "*" placed there.

@diox do you see any issue in trying to match even more the kind of checks that the patch on the Firefox side is going to be doing? is there any case covered by this method but not by the one in Firefox?

In the current version of that patch, the origin formatted does look like in the following snippet (pasting here because phabricator does not allow easily to link a certain like in a revision):
```
  origin(string, context) {
    let url;
    try {
      url = new URL(string);
    } catch (e) {
      throw new Error(`Invalid origin: ${string}`);
    }
    if (!/^https?:/.test(url.protocol)) {
      throw new Error(`Invalid origin must be http or https for URL ${string}`);
    }
    // url.origin is punycode so a direct check against string wont work.
    // url.href appends a slash even if not in the original string, we we
    // additionally check that string does not end in slash.
    if (string.endsWith("/") || url.href != new URL(url.origin).href) {
      throw new Error(
        `Invalid origin for URL ${string}, replace with origin ${url.origin}`
      );
    }
    if (!context.checkLoadURL(url.origin)) {
      throw new Error(`Access denied for URL ${url}`);
    }
    return url.origin;
  },
``` 

Besides the returned value (and the call to context.checkLoadURL, which it shouldn't matter nor be needed for the addons-linter validation check), it looks we may be able to derive the addons-linter validation function with a almost 1-1 match with the Firefox checks.

wdyt?

**Review by rpl — CHANGES_REQUESTED** [review]
I'm going to look more closely to the rest of the PR, but in the meantime I wanted to bring up a couple of comments I got based on the last review passes to the patch on the Firefox side that I've been doing today.

File: src/schema/updates/manifest.json
Comment (willdurand) [reply src/schema/updates/manifest.json:112]: Isn't the empty array only for MV2?

Comment (diox) [reply src/schema/updates/manifest.json:112]: That's correct, we now want the empty array allowed, this needs to be updated. Eventually, when we start enforcing, AMO will add an empty array to any submission that are missing the property (explicitly disallowing their install from any remote origin other than AMO).

File: src/schema/formats.js
Comment (diox) [reply src/schema/formats.js:118]: Yes, I haven't had the time to do that yet but I plan to.

Comment (diox) [reply src/schema/formats.js]: Did that in https://github.com/mozilla/addons-linter/pull/4018/commits/86a1a32e5a314e9a7f3c64b3a02a2bd529c36e7d (and https://github.com/mozilla/addons-linter/pull/4018/commits/104f477a3e7d81d19e595aaa9534ac8ed9d3746c to make `eslint` happy).

I kept the logic but had to make a few changes:
- Changed the `throw`s for `return false`
- Returned `true` instead of `url.origin` at the end
- Dropped the `context` variable and the `context.checkLoadURL` call
- Add something for the wildcards (which is more for convenience so that developers don't try to use them only to realize it doesn't work that way when they try to install their add-on later). 
- Renamed `string` to `value` for consistency with the rest of the file

File: src/schema/updates/manifest.json
Comment (diox) [reply src/schema/updates/manifest.json]: https://github.com/mozilla/addons-linter/pull/4018/commits/e83e4796e0f48aa302dfc0dd0deeabf5d419abd0

**Review by rpl — COMMENTED** [review]
This looks pretty good to me.

As a side note, I would really like if we could also assign a more specific rule code for invalid install_origins values, so that we can also make sure to show a message that would help the developer to know how to make the manifest key value valid, e.g. either:
- with a link to a doc page (as we do for permissions and optional permissions invalid values)
- or including the expected value in the error message when possible  as we are doing in the error we are going to log on the Firefox side).

To achieve that I think we may use a strategy similar to the one we use for the validation errors related to the `permissions` and `optional_permissions`, where the error got from the JSONSchema validator is then "changed / improved" in the ManifestJSONParser class (from the [erroLookup method here](https://github.com/mozilla/addons-linter/blob/ed8da2e1aedeceefc5f66d8911c111d699781a17/src/parsers/manifestjson.js#L222)).

Given that the changes to `ManifestJSONParser`'s `errorLookup` would require some more work (and potentially bring some more risks with it, depending from the amount and kind of changes we may have to apply to it), personally I would be ok to defer that to a separate follow up and proceed with this part as is in the short run. 

@willdurand wdyt?

File: src/schema/formats.js
Comment (willdurand) [inline src/schema/formats.js:110]: I don't know if our code supports it yet but theoretically we could use a catch without an error if we don't use it:

```suggestion
  } catch {
```

File: src/schema/formats.js
Comment (willdurand) [inline src/schema/formats.js:119]:
```suggestion
  // url.origin is punycode so a direct check against string won't work.
```

File: src/schema/formats.js
Comment (willdurand) [inline src/schema/formats.js:121]:
```suggestion
  // url.href appends a slash even if not in the original string, so we
  // additionally check that the value does not ends with slash.
```

File: tests/unit/schema/test.formats.js
Comment (willdurand) [inline tests/unit/schema/test.formats.js:133]: would be good to put actual descriptions/sentences here and in the other _added_ test cases, otherwise the output will look like:

```
it trailing slash
it domain with no path but with query string
```

**Review by willdurand — COMMENTED** [review]
some minor comments

**Comment by willdurand** [comment]
> Given that the changes to ManifestJSONParser's errorLookup would require some more work (and potentially bring some more risks with it, depending from the amount and kind of changes we may have to apply to it), personally I would be ok to defer that to a separate follow up and proceed with this part as is in the short run.
>
> @willdurand wdyt?

+1, I think we can merge this PR now (after some more cleanups if possible) and update the error message in a follow-up, which @diox would take care of ideally.

File: tests/unit/schema/test.formats.js
Comment (diox) [reply tests/unit/schema/test.formats.js:133]: I did this in the other file, but here I was trying to stay consistent on the other URL formats tests, plus those tests make several assertions (some true, some false) so I wasn't sure what to do...

File: src/schema/formats.js
Comment (diox) [reply src/schema/formats.js:118]: Yeah that surprised me, but tests caught it :)

File: tests/unit/schema/test.formats.js
Comment (willdurand) [reply tests/unit/schema/test.formats.js:133]: Yeah, it is quite involved actually so let's forget about this suggestion.

**Review by willdurand — APPROVED** [review]

**Review by rpl — APPROVED** [review]
lgtm too

I did also confirm that:
- by reimporting the schema data from beta, no changes are applied to the imported schema (in other words the install_origins definition injected from the updates schema is imported as is into the imported schema)
- by importing the schema data from mozilla-central tip, install_origins definition from mozilla-central gets the additional `maxItems: 5` from the schema updates as expected (and all tests added in this PR still pass successfully as expected)

**Pull request merged**
